Added ability to change the Kick Message displayed to players

keir-nellyer opened the pull request:
Added ability to change the Kick Message displayed to players when the server is shutdown and expanded the /stop command to support this feature.

Also added a temp shutdown kick message which sets the shutdown kick
message for that session, this is mainly used for the /stop
[kick-message] command.

When the server shuts down the temporary kick message will be shown if
it has been set, otherwise the one set in Bukkit.yml will be shown.

JIRA Ticket: https://bukkit.atlassian.net/browse/BUKKIT-3031
CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/937

Wolvereness commented on src/main/java/org/bukkit/command/defaults/StopCommand.java:
```diff
@@ -21,6 +23,10 @@ public StopCommand() {
         if (!testPermission(sender)) return true;
 
+        if (args.length > 0){
+            Bukkit.setTempShutdownKickMessage(ChatColor.translateAlternateColorCodes("&".charAt(0), StringUtils.join(args, " ")));
```
"&".charAt(0).... Never thought I'd see that suggestion come back...

keir-nellyer replied:
@Wolvereness Is there a better way of doing that?

md-5 replied:
Just '&'.....

keir-nellyer replied:
Thanks @md-5 never knew you could do that lol!

keir-nellyer commented:
@EdGruberman It works, as long as the & and the character code are not right beside each other http://imgur.com/iP0v1

Also for the temporal shutdown kick messages, it was tricky to do this across packages, as the StopCommand is in Bukkit and the code to kick the players at shutdown is in the NMS. So instead I made it into an API. Would there be a better way to do this? If so I will edit the commit.

keir-nellyer commented:
Removed the ChatColor translations

EdGruberman commented:
Ah, nice check on the single ampersand!  I forgot it was smart enough for that.  Still would have set a precedent to use that character so I'm glad to see it removed here.  Plugins could always use the API to allow color if desired, but by default no reason to set any arbitrary implementations.

And coincidence that I see you just updated this right now also. I think you removed your temporary message concept.  Using the same API for both seems okay to me, but instead I would be personally tempted to add another method such as Server.shutdown(String reason) myself.

keir-nellyer commented:
@EdGruberman Yes although to quote another comment I made on the CraftBukkit PR comments

"I would add it to the shutdown() method but it is not directly linked with the stop() method in MinecraftServer. When the shutdown() method in CraftServer is called it changes isRunning to false in MinecraftServer. Then the run() method is called in MinecraftServer (haven't traced this back yet) which calls the stop() method. The players are then kicked from the r() function in ServerConfigurationManagerAbstract. So basically what I'm saying is it is not as simple as passing a perimeter through a function.

Please let me know if I am wrong"

keir-nellyer commented:
@EdGruberman I suppose we could add a function kick the players in the CraftServer.shutdown() method, but would they be able to re-connect while the worlds are saving?

EdGruberman commented:
You are thinking too complex.  Simply store the message in a private field for CraftServer if it's passed in.  If it's null (from the original method call) then you use the default message specified in the bukkit.yml.  Keeps it all implementation specific, with a logical API.

keir-nellyer commented:
@EdGruberman How would I access the private/protected field as the stop/kick methods are in another package? Can we use reflection to access it?

EdGruberman commented:
Sorry, it would have to be public and probably make more sense to simply store it directly in ServerConfigurationManagerAbstract itself as that is what uses it.

Also, I responded in Bukkit/CraftBukkit#937 to try and consolidate this conversation.

keir-nellyer commented:
@EdGruberman I can't do that from the StopCommand in Bukkit though as we can only use the api in there.

keir-nellyer committed:
Added ability to change the Kick Message displayed to players when the
server is shutdown and expanded the /stop command to support this
feature.

Also added a temp shutdown kick message which sets the shutdown kick
message for that session, this is mainly used for the /stop
[kick-message] command.

When the server shuts down the temporary kick message will be shown if
it has been set, otherwise the one set in Bukkit.yml will be shown.

keir-nellyer commented:
See this comment for the latest update https://github.com/Bukkit/CraftBukkit/pull/937#issuecomment-10858462

feildmaster commented:
See craftbukkit https://github.com/Bukkit/CraftBukkit/pull/937#issuecomment-11039040
